File: java/src/main/java/de/elakito/escdd/ESCDDCodec.java
package de.elakito.escdd;

import java.io.ByteArrayOutputStream;
import java.util.regex.Pattern;

import static java.nio.charset.StandardCharsets.UTF_8;

public class ESCDDCodec {
    private Pattern validCharRE;
    private Pattern validFirstCharRE;
    private boolean uppercase;
    private char escChar;

    public ESCDDCodec(String validCharRE, String validFirstCharRE, char escChar) {
        this(validCharRE, validFirstCharRE, escChar, false);
    }

    public ESCDDCodec(String validCharRE, String validFirstCharRE, char escChar, boolean uppercase) {
        this.validCharRE = Pattern.compile(validCharRE);
        this.validFirstCharRE = Pattern.compile(validFirstCharRE);
        this.uppercase = uppercase;
        if (!this.validCharRE.matcher(String.valueOf(escChar)).find()) {
            throw new RuntimeException(String.format("%c not among the valid characters %s", escChar, validCharRE));
        }
        if (!this.validFirstCharRE.matcher(String.valueOf(escChar)).find()) {
            throw new RuntimeException(String.format("%c not among the valid first characters %s", escChar, validFirstCharRE));
        }
        this.escChar = escChar;
    }

    public String encode(String value) {
        StringBuffer buf = new StringBuffer();
        boolean escaped = false;

        for (int i = 0; i < value.length(); i++) {
            Pattern re = i == 0 ? validFirstCharRE : validCharRE;
            String c = value.substring(i, i+1);
            if (value.charAt(i) != escChar && re.matcher(c).find()) {
                buf.append(value.charAt(i));
            } else {
                for (byte b : c.getBytes(UTF_8)) {
                    buf.append(String.format(uppercase ? "%c%X" : "%c%x", escChar, b));
                }
                escaped = true;
            }
        }
        return escaped ? buf.toString() : value;
    }

    public String decode(String value) {
        StringBuilder buf = new StringBuilder();
        ByteArrayOutputStream bdec = new ByteArrayOutputStream();
        if (value.indexOf(escChar) >= 0) {
            for (int i = 0; i < value.length(); i++) {
                char c = value.charAt(i);
                if (c == escChar) {
                    try {
                        bdec.write(Integer.parseInt(value.substring(i + 1, i + 3), 16));
                        i += 2;
                    } catch (NumberFormatException e) {
                        //ignore
                    }
                } else {
                    if (bdec.size() > 0) {
                        buf.append(bdec.toString(UTF_8));
                        bdec.reset();
                    }
                    buf.append(c);
                }
            }
            return buf.toString();
        } else {
            return value;
        }
    }
}

File: scala/src/main/scala/ESCDDCodec.scala
import java.nio.charset.StandardCharsets
import java.nio.charset.StandardCharsets.UTF_8
import scala.collection.mutable.ArrayBuilder
import scala.io.Codec.UTF8

class ESCDDCodec(var validCharPattern: String, var validFirstCharPatter: String, var escChar: Char, var uppercase: Boolean = false) {
  private val validCharRE = validCharPattern.r
  private val validFirstCharRE = validFirstCharPatter.r

  require(validCharRE.findFirstIn(escChar.toString) != None, s"{escChar} not among the valid characters {validCharPattern}")
  require(validFirstCharRE.findFirstIn(escChar.toString) != None, s"{escChar} not among the valid first characters {validFirstCharPattern}")

  def encode(value: String): String = {
    val buf = new StringBuilder
    var escaped = false
    for (i <- 0 to value.length - 1) {
      val re = i match {
        case 0 => validFirstCharRE
        case _ => validCharRE
      }
      val c = value.substring(i, i+1)
      if (value.charAt(i) != escChar && re.findFirstIn(c) != None) {
        buf.append(c)
      } else {
        for (b <- c.getBytes(UTF_8)) {
          buf.append(if (uppercase) f"$escChar%c$b%X" else f"$escChar%c$b%x")
        }
        escaped = true
      }
    }
    escaped match {
      case true => buf.toString
      case _ => value
    }
  }

  def decode(value: String): String = {
    val buf = new StringBuilder
    var bdec = new ArrayBuilder.ofByte
    if (value.contains(escChar)) {
      var i = 0
      while (i < value.length) {
        val c = value.charAt(i)
        if (c == escChar) {
          bdec.addOne(Integer.parseInt(value.substring(i + 1, i + 3), 16).toByte)
          i += 2
        } else {
          if (bdec.length > 0) {
            buf.append(new String(bdec.result, StandardCharsets.UTF_8))
            bdec.clear
          }
           buf.append(c)
        }
        i += 1
      }
      return buf.toString
    } else {
      return value
    }
  }
}
